File: LeetCode/dp/LongestValidParentheses.java
package dp;

public class LongestValidParentheses {
    public int longestValidParentheses(String s) {
        int longest = 0;
        int dp[] = new int[s.length()];
        for (int i = 1; i < s.length(); i++) {
            if (s.charAt(i) == ')') {
                if (s.charAt(i - 1) == '(') {
                    dp[i] = (i >= 2 ? dp[i - 2] : 0) + 2;
                } else if (
                        i - dp[i - 1] - 1 >= 0 && s.charAt(i - dp[i - 1] - 1) == '('
                ) {
                    dp[i] = dp[i - 1] +
                            ((i - dp[i - 1]) >= 2 ? dp[i - dp[i - 1] - 2] : 0) +
                            2;
                }
                longest = Math.max(longest, dp[i]);
            }
        }
        return longest;
    }

    public static void main(String[] args) {
        LongestValidParentheses solution = new LongestValidParentheses();
        String s = "(()))())(";
        System.out.println(solution.longestValidParentheses(s)); // Expected output: 4
    }
}

File: LeetCode_Python/stack/LongestValidParentheses.py
class LongestValidParentheses:
    def longestValidParentheses(self, s: str) -> int:
        stack = [-1]
        max_length = 0

        for i in range(len(s)):
            char = s[i]
            if char == "(":
                stack.append(i)
            else:
                stack.pop()
                if not stack:
                    stack.append(i)
                else:
                    max_length = max(max_length, i - stack[-1])

        return max_length


# Time Complexity: O(n)
# Space Complexity: O(n)

if __name__ == "__main__":
    solution = LongestValidParentheses()

    print("Expected Output: 2")
    print("Actual Output:", solution.longestValidParentheses("(()"))

    print("Expected Output: 4")
    print("Actual Output:", solution.longestValidParentheses(")()())"))
